osdep/clipboard.h

```diff
@@ -0,0 +1,72 @@
+/*
+ * Clipboard access routines
+ *
+ * This file is part of mpv.
+ *
+ * mpv is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * mpv is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with mpv.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "player/core.h"
+#include "video/mp_image.h"
+
+enum m_clipboard_type {
+    CLIPBOARD_UNKNOWN,
+    CLIPBOARD_TEXT,
+    CLIPBOARD_URL,
+    CLIPBOARD_PATH,
+    CLIPBOARD_IMAGE,
+};
+
+struct m_clipboard_item {
+    enum m_clipboard_type type;
+    union {
+        char *string;
+        struct mp_image *image;
+    };
+};
+
+struct clipboard_state {
+#if HAVE_COCOA
+    long changeCount;
+#endif
+};
+
+enum m_clipboard_return {
+    CLIPBOARD_OK = 1,
+    CLIPBOARD_NONE = 0,
+    CLIPBOARD_FAILED = -1,
+};
+
+int m_clipboard_get(struct MPContext *ctx, struct m_clipboard_item *item);
+int m_clipboard_set(struct MPContext *ctx, const struct m_clipboard_item *item);
+bool m_clipboard_poll(struct MPContext *ctx);
```

Comment:
Re: context, I currently have `struct clipboard_state` inited at startup and owned by `MPContext`, and it'd be easy enough to add an explicit `uninit` at shutdown. I don't think that needs to be added in this PR, though, since nothing in it precludes adding explicit shutdown function later.

Re: X/Wayland, I think we'd want to either have both run continuously (can they run in a single thread, `select()`ing both?) and return data from whichever was updated most recently, or have a user setting for which to use (defaulting to whichever we'd use for vo).

Reply:
> Personally I would like to limit clipboard to 1 backend at a time, to keep it sane.

Agreed.

I think a reasonable way would be to delegate/marry this to a VO. Then the event loop can be reused to handle the necessary clipboard protocol.
Then add a way to have a global clipboard manager for the sake of win32 and mac (where this doesn't depend on a VO).

Reply:
As far as I can tell, the win32 clipboard can be read and written without touching a window at all, and `GetClipboardSequenceNumber()` should allow for an implementation of `m_clipboard_poll` that works much like the macOS one.

I don't understand enough about the relevant X/Wayland APIs to be able to confidently say much about how they'll plug into any of this, or to be able to design a vo-based API to accommodate them.

Reply:
Ah, hmm, I'd missed those caveats; `m_clipboard_poll` is called on every render loop iteration, so that's probably not the best solution.

Reply:
Alternately, the vo could monitor clipboard events and fire `mp_notify_property` when applicable (might need a dispatch call?), and `m_clipboard_poll` could remain a no-op on windows. Not sure if any of this actually implies changes to the API as implemented here at all, then; the vo can just reach into `mpctx->clipboard` to hand it a `HWND` to pass to `OpenClipboard` when performing read/write calls.